src/kde/kde.cc

```diff
@@ -30,124 +30,437 @@ namespace rtt_kde {
+    // ZERO IS ZERO AND THE LOG TRANSFORM CAN MAKE THE ZEROS NOT MATCH... SO FIX IT LIKE THIS
+    if (rtt_dsxx::soft_equiv(result[i], 0.0) && rtt_dsxx::soft_equiv(distribution[i], 0.0))
+      result[i] = distribution[i];
   }
 
-  double reconstruction_conservation = std::accumulate(result.begin(), result.end(), 0.0);
+  return result;
+}
+
+//------------------------------------------------------------------------------------------------//
+/*!
+ * \brief KDE apply conservation
+ * 
+ * \pre Apply conservation fix to the new distribution so
+ * sum(original_distribution) == sum(new_distribution)
+ *
+ * \param[in] original_distribution original data to be reconstructed
+ * \param[in,out] new_distribution original data to be reconstructed
+ * \param[in] domain_decomposed bool
+ *
+ */
+void kde::apply_conservation(const std::vector<double> &original_distribution,
+                             std::vector<double> &new_distribution,
+                             const bool domain_decomposed) const {
+
+  const size_t local_size = original_distribution.size();
+  Insist(new_distribution.size() == local_size,
+         "Original and new distributions must be the same size");
+
+  // compute absolute solution
+  std::vector<double> abs_distribution(local_size, 0.0);
+  for (size_t i = 0; i < local_size; i++) {
+    if (!rtt_dsxx::soft_equiv(new_distribution[i], original_distribution[i], 1e-12))
+      abs_distribution[i] = fabs(new_distribution[i]);
+  }
+
+  // compute totals to be used in residual calculation
+  double original_conservation =
+      std::accumulate(original_distribution.begin(), original_distribution.end(), 0.0);
+  double reconstruction_conservation =
+      std::accumulate(new_distribution.begin(), new_distribution.end(), 0.0);
+  double abs_distribution_conservation =
+      std::accumulate(abs_distribution.begin(), abs_distribution.end(), 0.0);
 
   if (domain_decomposed) {
     // accumulate global contribution
+    rtt_c4::global_sum(original_conservation);
     rtt_c4::global_sum(reconstruction_conservation);
+    rtt_c4::global_sum(abs_distribution_conservation);
   }
 
-  if (!rtt_dsxx::soft_equiv(reconstruction_conservation, 0.0) &&
-      !rtt_dsxx::soft_equiv(global_conservation, 0.0)) {
-    // Totals are non-zero so scale the result for conservation
-    for (int i = 0; i < local_size; i++)
-      result[i] *= global_conservation / reconstruction_conservation;
-  } else {
-    // a zero distribution is possible. If it occurs fall back to residual conservation;
-    const double res = global_conservation - reconstruction_conservation;
-    for (int i = 0; i < local_size; i++)
-      result[i] += res / double(size);
+  // Apply residual
+  if (abs_distribution_conservation > 0.0) {
+    const double res = original_conservation - reconstruction_conservation;
+    for (size_t i = 0; i < local_size; i++)
+      new_distribution[i] += res * abs_distribution[i] / abs_distribution_conservation;
```

Comment:
Any risk of division by zero?

Reply:
Not here because it is protected by the if greater then zero criteria.